File: core/src/main/java/ru/surfstudio/android/core/ui/base/navigation/activity/route/ActivityWithResultRoute.java
package ru.surfstudio.android.core.ui.base.navigation.activity.route;


import android.content.Intent;

import java.io.Serializable;

import ru.surfstudio.android.core.ui.base.screen.event.result.SupportOnActivityResultRoute;

/**
 * см {@link ActivityRoute}
 * @param <T> тип результата
 */
public abstract class ActivityWithResultRoute<T extends Serializable> extends ActivityRoute
        implements SupportOnActivityResultRoute<T> {

    private static final int MAX_REQUEST_CODE = 32768;

    @Override
    public Intent prepareResultIntent(T resultData){
        Intent i = new Intent();
        i.putExtra(EXTRA_RESULT, resultData);
        return i;
    }

    @Override
    public T parseResultIntent(Intent resultIntent){
        return (T)resultIntent.getSerializableExtra(EXTRA_RESULT);
    }

    @Override
    public final int getRequestCode(){
        return Math.abs(this.getClass().getCanonicalName().hashCode() % MAX_REQUEST_CODE);
    }

}

File: navigation/lib-navigation-observer/src/main/java/ru/surfstudio/android/navigation/observer/route/ActivityWithResultRoute.kt
package ru.surfstudio.android.navigation.observer.route

import android.content.Intent
import ru.surfstudio.android.navigation.route.activity.ActivityRoute
import java.io.Serializable

/**
 * Route for activity with result [ActivityRoute]
 * You should use this route when launching system screens.
 *
 * @param <T> type of the result returned from activity
 */
abstract class ActivityWithResultRoute<T : Serializable> : ActivityRoute(), ResultRoute<T> {

    /**
     * Screen identifier, used to avoid collisions where one result can be returned to many subscribers.
     * Result of the activity will be returned only to subscribers which uniqueId is the same.
     */
    abstract val uniqueId: String

    override fun getId(): String {
        return "${this.javaClass.canonicalName}$uniqueId"
    }

    /**
     * Parsing result Intent to given type.
     */
    abstract fun parseResultIntent(resultCode: Int, resultIntent: Intent?): T
}